teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+   runAfter: get-source
+   steps:
+  - name: unit-test
+    image: docker.io/library/golang:$(params.version)
+    script: <script here>
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   foobarRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+finally:
+- name: update-slack
+  params:
+  - name: message
+    value: "Tests completed with $(tasks.run-unit-tests.status) status"
+```
+
+We will need to add something to indicate how to schedule Foobars now that we won't have the convience of drawing the
+line around Tasks; we could combine this idea with one of the others in this proposal.
+
+(@ImJasonH has some ideas here that are much more aligned with generics in other langs, he might want to update the
+example here.)
+
+Pros:
+* Maybe the distinction between Tasks and Pipelines has just been making things harder for us
+* Maybe this is a natural progression of the "embedding" we already allow in pipelines?
+
+Cons:
+* Pretty dramatic API change
+* Foobars can contain Foobars can contain Foobars can contain Foobars
+
+### Custom Pipeline
+
+In this approach we solve this problem by making a Custom Task that can run a Pipeline using whatever scheduling
+mechanism is preferred; this assumes the custom task is the ONLY way we support a different scheduling than
+Task to pod going forward (even if we pick a different solution it could make sense to implement it as a custom task
+first).
+
+Pros:
+* Doesn't change anything about our API
+
+Cons:
+* If this custom task is widely adopted, could fork our user community
+
+### Create a new Grouping CRD
+
+In this approach we create a new CRD, e.g. `TaskGroup` that can be used to group Tasks together. Pipelines can refer to
+TaskGroups, and they can even embed them.
+
+For example:
+
+```yaml
+kind: TaskGroup
+metadata:
+  name: build-test-deploy
+spec:
+ workspaces:
+  - name: source-code
+  - name: test-results
+ tasks:
+ - name: get-source
+   workspaces:
+   - name: source-code
+     workspace: source-code
+   taskRef:
+     name: git-clone
+   params:
+   - name: url
+      value: $(params.url)
+    - name: revision
+      value: $(params.revision)
+ - name: run-unit-tests
+   runAfter: get-source
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   taskRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+```
+
+We could decide if we only support sequential execution, or support an entire DAG. Maybe even finally?
+
+Cons:
```

Comment:
No pros at all?

A task group CRD could be used do define a number of different execution strategies:

- sequential only
- switch (only one based on input)
- race (finish as soon as at least one is finished)

This is a different kind of problem, but different strategies may fit better with specific scheduling approaches.

Reply:
ive added this as a pro - kinda sounds to me like you're describing custom tasks in general tho? (e.g. you can have a switch statement custom task)